torch/onnx/_internal/diagnostics/_diagnostic.py

```diff
@@ -0,0 +1,92 @@
+"""Diagnostic components for PyTorch ONNX export."""
+
+from typing import Any, Optional, Tuple
+
+import torch
+from torch.onnx._internal.diagnostics import _rules, infra
+
+levels = infra.Level
+
+
+class ExportDiagnostic(infra.Diagnostic):
+    """Base class for all export diagnostics.
+
+    This class is used to represent all export diagnostics. It is a subclass of
+    infra.Diagnostic, and adds additional methods to add more information to the
+    diagnostic.
+    """
+
+    def __init__(
+        self,
+        rule: infra.Rule,
+        level: infra.Level,
+        message_args: Optional[Tuple[Any, ...]],
+        **kwargs,
+    ) -> None:
+        super().__init__(rule, level, message_args, **kwargs)
+
+    def with_cpp_stack(self) -> "ExportDiagnostic":
```

Comment:
`from typing_extension import Self`

```suggestion
    def with_cpp_stack(self) -> Self:
```